### ISSUE TITLE
Header Content-Length is controlled by the engine and cannot be set explicitly

### ISSUE THREAD

**Issue by bartwinski** [ktorio/ktor#1993]
When using the Ktor http client (ktor 1.3.2) for PUT request without payload the **Content-Length** is not set, but it should be.

But, setting the **Content-Length** explicitly in the code with "**_header(HttpHeaders.ContentLength, 0)_**" cause the following exception:

```
io.ktor.http.UnsafeHeaderException: Header Content-Length is controlled by the engine and cannot be set explicitly
	at io.ktor.client.engine.HttpClientEngineKt.validateHeaders(HttpClientEngine.kt:140)
	at io.ktor.client.engine.HttpClientEngineKt.access$validateHeaders(HttpClientEngine.kt:1)
	at io.ktor.client.engine.HttpClientEngine$install$1.invokeSuspend(HttpClientEngine.kt:62)
	at io.ktor.client.engine.HttpClientEngine$install$1.invoke(HttpClientEngine.kt)
	at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:318)
	at io.ktor.util.pipeline.SuspendFunctionGun.proceed(PipelineContext.kt:163)
	at io.ktor.util.pipeline.SuspendFunctionGun.execute(PipelineContext.kt:183)
	at io.ktor.util.pipeline.Pipeline.execute(Pipeline.kt:27)
	at io.ktor.client.features.HttpSend$DefaultSender.execute(HttpSend.kt:115)
	at io.ktor.client.features.HttpSend$Feature$install$1.invokeSuspend(HttpSend.kt:79)
	at io.ktor.client.features.HttpSend$Feature$install$1.invoke(HttpSend.kt)
	at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:318)
	at io.ktor.util.pipeline.SuspendFunctionGun.proceed(PipelineContext.kt:163)
	at io.ktor.client.features.HttpRequestLifecycle$Feature$install$1.invokeSuspend(HttpRequestLifecycle.kt:34)
	at io.ktor.client.features.HttpRequestLifecycle$Feature$install$1.invoke(HttpRequestLifecycle.kt)
	at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:318)
	at io.ktor.util.pipeline.SuspendFunctionGun.proceed(PipelineContext.kt:163)
	at io.ktor.util.pipeline.SuspendFunctionGun.execute(PipelineContext.kt:183)
	at io.ktor.util.pipeline.Pipeline.execute(Pipeline.kt:27)
	at io.ktor.client.HttpClient.execute(HttpClient.kt:164)
	at io.ktor.client.statement.HttpStatement.executeUnsafe(HttpStatement.kt:104)
	at io.ktor.client.statement.HttpStatement.execute(HttpStatement.kt:43)
	at io.ktor.client.statement.HttpStatement.execute(HttpStatement.kt:58)
```

**Comment by Stexxe** [comment]
Could you please justify why this header should be set?
For example, when I send `PUT` request via `curl` (Linux utility) like this: `curl -X PUT http://localhost:8080/` it doesn't add `Content-Length` header. It adds this header only if there is some data (`curl -X PUT -d arg=val http://localhost:8080/`).

**Comment by bartwinski** [comment]
Content-length must  be specify in a POST or PUT request, or they must use transfer-encoding:chunked. One of content-length or transfer-encoding:chunked should be present. For more see the HTTP spec, section 4.4.
https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.4
If there is no content-length specified, how the server should know how much payload data are available ?
Your curl works without content-length , because the server side is so implemented, that it ignores.

**Comment by e5l** [comment]
Hey @bartwinski, could you tell me how you setting the body? 
You can configure `contentType` header in `OutgoingContent` body.

**Comment by bartwinski** [comment]
I am not setting the body at all, because there is no body needed for this particular PUT request.

This is how I currently call it
`httpClient.put<HttpResponse>(url) {
   header(X_API_KEY, "some api key")
}`

the following workaround would work ... then the Content-Length is set by the http client to "0" and the server is happy
`httpClient.put<HttpResponse>(url) {
  header(X_API_KEY, "some api key")
  body = TextContent("", ContentType.Application.Json)
}`

but why should I set an empty body if there is no body needed for my particular PUT request.
So if the API definition of the service on the server side doesn't define the need for the payload, why should I set it.
